### PR TITLE
Fix potential mana rounding

### PR DESCRIPTION
**Pull request by oliviasaa**
Solves issues:
- https://github.com/iotaledger/iota-core/issues/391
- https://github.com/iotaledger/iota-core/issues/366
- https://github.com/iotaledger/iota-core/issues/363

### CODE REVIEW

File: mana.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func (r RewardsParameters) TargetReward(index EpochIndex, api API) (Mana, error)
 		return Mana(api.ComputedFinalReward()), nil
 	}
 
-	decayedInitialReward, err := api.ManaDecayProvider().RewardsWithDecay(Mana(api.ComputedInitialReward()), index, index)
+	decayedInitialReward, err := api.ManaDecayProvider().RewardsWithDecay(Mana(api.ComputedInitialReward()), 0, index)
```
Comment (oliviasaa) [inline mana.go]: it starts in 0 or 1?

File: mana_decay_provider.go [hunk 2]
```diff
@@ -1,6 +1,9 @@
 package iotago
 
 import (
+	"fmt"
```
Comment (oliviasaa) [inline mana_decay_provider.go]:
```suggestion
```

File: mana_decay_provider.go [hunk 3]
```diff
@@ -188,14 +196,20 @@ func (p *ManaDecayProvider) ManaGenerationWithDecay(amount BaseToken, creationSl
...
 		// result = potentialMana_0 - potentialMana_n_1 + potentialMana_n
 		//nolint:golint,revive,nosnakecase,stylecheck // taken from the formula, lets keep it that way
 		result, err := safemath.SafeSub(potentialMana_0, potentialMana_n_1)
 		if err != nil {
 			return 0, err
 		}
-
+		printVar, _ := safemath.SafeAdd(result, potentialMana_n)
 		//nolint:golint,revive,nosnakecase,stylecheck // taken from the formula, lets keep it that way
+		fmt.Println(p.LowerBoundPotentialMana(amount, creationSlot, targetSlot), printVar, p.UpperBoundPotentialMana(amount, creationSlot, targetSlot))
```
Comment (oliviasaa) [inline mana_decay_provider.go]:
```suggestion
```

File: mana_decay_provider.go [hunk 4]
```diff
@@ -1,6 +1,9 @@
 package iotago
 
 import (
+	"fmt"
+	"math"
```
Comment (oliviasaa) [inline mana_decay_provider.go]: this package is used just for the function I'm supposed to move to the test file

File: mana_decay_provider.go [hunk 5]
```diff
@@ -188,14 +196,20 @@ func (p *ManaDecayProvider) ManaGenerationWithDecay(amount BaseToken, creationSl
...
+		}
+
 		// result = potentialMana_0 - potentialMana_n_1 + potentialMana_n
 		//nolint:golint,revive,nosnakecase,stylecheck // taken from the formula, lets keep it that way
 		result, err := safemath.SafeSub(potentialMana_0, potentialMana_n_1)
 		if err != nil {
 			return 0, err
 		}
-
+		printVar, _ := safemath.SafeAdd(result, potentialMana_n)
```
Comment (oliviasaa) [inline mana_decay_provider.go]:
```suggestion
```

File: mana_decay_provider.go [hunk 6]
```diff
@@ -41,9 +44,14 @@ func fixedPointMultiplication32Splitted(valueHi uint64, valueLo uint64, factor u
...
 func fixedPointMultiplication32(value uint64, factor uint64, scale uint64) uint64 {
+	extraShift := uint64(0)
+	if scale > 32 {
+		extraShift = scale - 32
+		scale = 32
+	}
 	valueHi, valueLo := splitUint64(value)
 
-	return mergeUint64(fixedPointMultiplication32Splitted(valueHi, valueLo, factor, scale))
+	return mergeUint64(fixedPointMultiplication32Splitted(valueHi, valueLo, factor, scale)) >> extraShift
```
Comment (oliviasaa) [inline mana_decay_provider.go]: here or in the other function?

File: mana_decay_provider.go [hunk 7]
```diff
@@ -208,3 +222,44 @@ func (p *ManaDecayProvider) RewardsWithDecay(rewards Mana, rewardEpoch EpochInde
...
+		slotsPerEpochFloat := math.Pow(2, 13)
+		constant := math.Exp(-delta/3) * (1 - math.Exp(-(epochDiffFloat-1)*delta/3)) / (1 - math.Exp(-delta/3))
+		potentialMana_n := float64(slotsBeforeNextEpoch) * floatAmount * floatGenRate * math.Exp(-epochDiffFloat*delta/3)
+		potentialMana_n_1 := constant * floatAmount * floatGenRate * slotsPerEpochFloat
+		potentialMana_0 := float64(slotsSinceEpochStart) * floatAmount * floatGenRate
+		return potentialMana_n + potentialMana_n_1 + potentialMana_0
+	}
+}
+
+func (p *ManaDecayProvider) LowerBoundPotentialMana(amount BaseToken, creationSlot SlotIndex, targetSlot SlotIndex) float64 {
```
Comment (oliviasaa) [inline mana_decay_provider.go]: explain where these bounds came from

File: mana.go (hunk 1 above)
Comment (oliviasaa) [reply mana.go]: 1

Comment (oliviasaa) [reply mana.go]: Epoch Index

File: mana_decay_provider.go [hunk 8]
```diff
@@ -157,8 +165,8 @@ func (p *ManaDecayProvider) ManaGenerationWithDecay(amount BaseToken, creationSl
 	creationEpoch := p.timeProvider.EpochFromSlot(creationSlot)
```
Comment (oliviasaa) [inline mana_decay_provider.go]: move this to after the if

File: mana_decay_provider.go [hunk 9]
```diff
@@ -174,7 +182,7 @@ func (p *ManaDecayProvider) ManaGenerationWithDecay(amount BaseToken, creationSl
 		return safemath.SafeAdd(manaDecayed, manaGenerated)
 
 	default:
-		c := Mana(fixedPointMultiplication32(uint64(amount), p.decayFactorEpochsSum, p.decayFactorEpochsSumExponent+p.generationRateExponent-p.slotsPerEpochExponent))
+		c := Mana(fixedPointMultiplication32(uint64(amount), p.decayFactorEpochsSum*p.generationRate, p.decayFactorEpochsSumExponent+p.generationRateExponent-p.slotsPerEpochExponent))
```
Comment (oliviasaa) [inline mana_decay_provider.go]: explain what's the difference: no decay if it's zero etcetc

File: mana_decay_provider.go (hunk 6 above)
Comment (oliviasaa) [reply mana_decay_provider.go]: check with the one above is used anywhere

Comment (oliviasaa) [reply mana_decay_provider.go]: merge them if is not used anywhere

File: mana_decay_provider.go [hunk 10]
```diff
@@ -195,6 +196,8 @@ func (p *ManaDecayProvider) ManaGenerationWithDecay(amount BaseToken, creationSl
 
 		return result, nil
 
+	// case 0 means that the creationSlot and targetSlot belong to subsequent epochs.
```
Comment (muXxer) [inline mana_decay_provider.go]:
```suggestion
	// case 1 means that the creationSlot and targetSlot belong to subsequent epochs.
```

**Review by muXxer — CHANGES_REQUESTED** [review]

File: mana_decay_provider_test.go [hunk 11]
```diff
@@ -368,3 +390,62 @@ func TestManaDecay_Rewards(t *testing.T) {
...
+		constant := math.Exp(-delta/3) * (1 - math.Exp(-(epochDiffFloat-1)*delta/3)) / (1 - math.Exp(-delta/3))
+		potentialMana_n := float64(slotsBeforeNextEpoch) * floatAmount * floatGenerationRate * math.Exp(-epochDiffFloat*delta/3)
+		potentialMana_n_1 := constant * floatAmount * floatGenerationRate * slotsPerEpochFloat
+		potentialMana_0 := float64(slotsSinceEpochStart) * floatAmount * floatGenerationRate
+
+		return potentialMana_n + potentialMana_n_1 + potentialMana_0
+	}
+}
+
+func (p *TestFloatManaDecayProvider) LowerBoundPotentialMana(amount iotago.BaseToken, creationSlot iotago.SlotIndex, targetSlot iotago.SlotIndex) float64 {
```
Comment (muXxer) [inline mana_decay_provider_test.go:443]: You added a review comment yourself before I moved that code:
`explain where these bounds came from`
Is this still a thing?

Comment (oliviasaa) [reply mana_decay_provider_test.go:443]: let's leave it like it is

**Review by muXxer — APPROVED** [review]

**Pull request merged**